### PR TITLE
Grappler: don't rewrite ArgMin/ArgMax over non-strictly-monotonic ops

### LINKED ISSUE #116436: Grappler ArithmeticOptimizer incorrectly strips non-strictly-monotonic ops from ArgMin/ArgMax
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

TensorFlow 2.21.0

### Custom code

Yes

### OS platform and distribution

Ubuntu Linux

### Mobile device

_No response_

### Python version

Python 3.11

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

CUDA 12 / cuDNN 9

### GPU model and memory

_No response_

## Summary

The `OptimizeMaxOrMinOfMonotonic` stage in Grappler's `ArithmeticOptimizer` transforms `ArgMin(f(x))` into `ArgMin(x)` for ops classified as "monotonic non-decreasing" by `IsElementWiseMonotonic()` in `op_types.cc`. This is incorrect for non-strictly-monotonic functions (e.g., `Relu`, `Relu6`, `Floor`, `Ceil`, `Rint`, `Sign`), which have flat regions that create ties and alter the argmin result.

## Minimal Reproducer

```python
import tensorflow as tf

x = tf.constant([[1.0, -1.0, -5.0, 2.0]])
# relu(x) = [1, 0, 0, 2] → argmin = 1 (first zero)
# x       = [1, -1, -5, 2] → argmin = 2 (most negative)

print("eager:", tf.argmin(tf.nn.relu(x), axis=-1).numpy())  # [1] ✓

tf.config.optimizer.set_experimental_options({"arithmetic_optimization": True})
print("arith ON: ", tf.function(lambda x: tf.argmin(tf.nn.relu(x), axis=-1))(x).numpy())  # [2] ✗

tf.config.optimizer.set_experimental_options({"arithmetic_optimization": False})
print("arith OFF:", tf.function(lambda x: tf.argmin(tf.nn.relu(x), axis=-1))(x).numpy())  # [1] ✓
```

**Output (TF 2.21.0, CUDA 12, Linux):**
```
eager:    [1]
arith ON:  [2]
arith OFF: [1]
```

## Affected ops

The following ops from `IsElementWiseMonotonic()` (in `tensorflow/core/grappler/op_types.cc`) are non-strictly-monotonic and confirmed to trigger incorrect results:

| Op | Flat region | Confirmed |
|----|------------|-----------|
| `Relu` | f(x)=0 for x≤0 | ✅ |
| `Relu6` | f(x)=0 for x≤0, f(x)=6 for x≥6 | ✅ |
| `Floor` | f(x) constant on [n, n+1) | ✅ |
| `Ceil` | f(x) constant on (n-1, n] | ✅ |
| `Rint` | f(x) constant on [n-0.5, n+0.5) | ✅ (same root cause as Floor) |
| `Sign` | f(x) constant on (-∞,0), {0}, (0,+∞) | Not tested, same principle |

The remaining ops in the list (Exp, Log, Sigmoid, Tanh, Softplus, etc.) are strictly monotonic and the optimization is valid for them.

## Root cause

The transformation `ArgMin(f(x)) → ArgMin(x)` is only valid when f is **strictly** monotonic, i.e., `a < b ⟹ f(a) < f(b)`. For non-strictly-monotonic functions like Relu, there exist distinct values a ≠ b where f(a) = f(b). When this happens, `ArgMin(f(x))` returns the first index among the tied values (per TF's tie-breaking rule), while `ArgMin(x)` returns the index of the true minimum — a different index.

## Suggested fix

In `op_types.cc`, split `IsElementWiseMonotonic` into strict and non-strict variants, or add a `bool* is_strictly_monotonic` out-parameter. Then in the optimizer stage, only apply the `ArgMin`/`ArgMax` transformation for strictly monotonic ops.

Non-strictly-monotonic ops to exclude: `Relu`, `Relu6`, `Floor`, `Ceil`, `Rint`, `Sign`.

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

x = tf.constant([[1.0, -1.0, -5.0, 2.0]])
# relu(x) = [1, 0, 0, 2] → argmin = 1 (first zero)
# x       = [1, -1, -5, 2] → argmin = 2 (most negative)

print("eager:", tf.argmin(tf.nn.relu(x), axis=-1).numpy())  # [1] ✓

tf.config.optimizer.set_experimental_options({"arithmetic_optimization": True})
print("arith ON: ", tf.function(lambda x: tf.argmin(tf.nn.relu(x), axis=-1))(x).numpy())  # [2] ✗

tf.config.optimizer.set_experimental_options({"arithmetic_optimization": False})
print("arith OFF:", tf.function(lambda x: tf.argmin(tf.nn.relu(x), axis=-1))(x).numpy())  # [1] ✓


**Output (TF 2.21.0, CUDA 12, Linux):**

eager:    [1]
arith ON:  [2]
arith OFF: [1]
```

### Relevant log output

```shell

```

### PR DESCRIPTION
**Pull request by zxuhan**
## Summary

- `OptimizeMaxOrMinOfMonotonicStage` rewrites `ArgMin(f(x)) -> ArgMin(x)` for any op tagged monotonic, but the rewrite is only valid when `f` is **strictly** monotonic. Ops like `Relu`, `Relu6`, `Floor`, `Ceil`, `Rint`, and `Sign` have flat regions: their ties resolve to a different index than the underlying input, so the rewrite returns the wrong index.
- `IsElementWiseMonotonic` now exposes an optional `is_strict` out-parameter that flags the non-strict subset. The optimizer uses it to skip the rewrite when the reduction is `ArgMin`/`ArgMax` and `f` is non-strict. Value-reduction rewrites (`Max`/`Min`/`MaxPool`) are unchanged because they only depend on the reduced value, not the index.

Fixes #116436.

## Test plan

Relying on internal CI to run the added test:

- `bazel test //tensorflow/core/grappler/optimizers:arithmetic_optimizer_test` — added `OptimizeArgMaxOrArgMinOfMonotonicElementWiseDoNotChangeNonStrict` covering `ArgMin(Relu(x))`; existing `OptimizeArgMaxOrArgMinOfMonotonicElementWise` (Sqrt — strict) should still pass.
- Repro from the issue: `tf.argmin(tf.nn.relu(x), axis=-1)` with `arithmetic_optimization=True` should return the same index as eager mode.

### CODE REVIEW

File: tensorflow/core/grappler/optimizers/arithmetic_optimizer_test.cc [hunk 1]
```diff
@@ -4042,6 +4042,37 @@ TEST_F(ArithmeticOptimizerTest, OptimizeArgMaxOrArgMinOfMonotonicElementWise) {
...
+  VerifyGraphsMatch(item.graph, output, __LINE__);
+
+  const auto tensors = EvaluateNodes(output, item.fetch);
+  ASSERT_EQ(tensors.size(), 1);
+  test::ExpectTensorEqual<int64_t>(tensors[0], tensors_expected[0]);
+}
+
 TEST_F(ArithmeticOptimizerTest,
        OptimizeMaxOrMinOfMonotonicElementWiseDoNotChangeFetchNode) {
   tensorflow::Scope s = tensorflow::Scope::NewRootScope();
```
Comment (CodersAcademy006) [inline tensorflow/core/grappler/optimizers/arithmetic_optimizer_test.cc:4078]: The additional `is_strict` gating makes sense to me for preserving ArgMin/ArgMax tie semantics. One edge case that may still be worth thinking about separately is numerically “strict” ops whose floating-point implementations can still collapse distinct values due to saturation/precision effects under certain dtypes/backend lowerings (similar to #118675).

Comment (CodersAcademy006) [inline tensorflow/core/grappler/optimizers/arithmetic_optimizer_test.cc:4078]: Good regression coverage here. I especially like that the test validates graph equivalence preservation in addition to the runtime output, since this optimization bug was fundamentally about unsafe rewrite legality rather than just a numerical mismatch.

File: tensorflow/core/grappler/op_types.cc [hunk 2]
```diff
@@ -225,8 +225,12 @@ bool IsDivNoNan(const NodeDef& node) { return node.op() == "DivNoNan"; }
 // Returns true if node represents a unary elementwise function that is
 // monotonic. If *is_non_decreasing is true, the function is non-decreasing,
 // e.g. sqrt, exp. *is_non_decreasing is false, the function is non-increasing,
-// e.g. inv.
-bool IsElementWiseMonotonic(const NodeDef& node, bool* is_non_decreasing) {
+// e.g. inv. If *is_strict is true, the function is strictly monotonic
+// (a < b => f(a) < f(b) for non-decreasing, or f(a) > f(b) for
+// non-increasing). Ops with flat regions (e.g. Relu, Floor) are monotonic
+// but not strictly monotonic.
+bool IsElementWiseMonotonic(const NodeDef& node, bool* is_non_decreasing,
```
Comment (CodersAcademy006) [inline tensorflow/core/grappler/op_types.cc:232]: I like the separation between monotonic and strictly-monotonic semantics here. One thing I’m wondering about though is whether the current classification risks becoming fragile over time as additional ops get added manually to these sets.

Would it make sense longer term to make the “strictness” property more explicit/documented at the optimizer-contract level somewhere, since transformations like ArgMin/ArgMax are correctness-sensitive to tie semantics rather than monotonicity alone?

**Review by CodersAcademy006 — COMMENTED** [review]
This looks like a solid fix overall. The important distinction to me is that the original optimization was relying on monotonicity as a sufficient condition for ArgMin/ArgMax rewrite legality, when the actual correctness boundary is preservation of ordering *and* tie structure. Separating strict vs non-strict monotonic semantics makes the optimizer contract substantially clearer and avoids a pretty subtle class of index-selection bugs.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thanks for working on this! This is an excellent catch. 
  
  Since index-based reductions like `ArgMin` / `ArgMax` depend on the first occurrence of minimum/maximum values, stripping out monotonic transformations with flat regions (like `Relu` or `Floor`) causes correctness mismatches compared to eager execution. The approach to restrict this optimization to strictly monotonic functions is correct.
  
  Before we can merge this, could you please update the PR with the following comments and tests?
  
  ### 1. Document the new parameter
  Please update the class documentation for `IsElementWiseMonotonic` in `tensorflow/core/grappler/op_types.h` and `tensorflow/core/grappler/op_types.cc` to reflect the new `is_strict` parameter:
  
  ```cpp
  // Returns true if node represents a unary elementwise function that is
  // monotonic. If *is_non_decreasing is true, the function is non-decreasing,
  // e.g. sqrt, exp. *is_non_decreasing is false, the function is non-increasing,
  // e.g. inv.
  // If is_strict is provided, sets *is_strict to true if the function is strictly
  // monotonic, and false if it contains flat regions (e.g. Relu, Floor).
  bool IsElementWiseMonotonic(const NodeDef& node, bool* is_non_decreasing, bool* is_strict = nullptr);
  
  ```
**2. Add an explicit regression test**
Please add a unit test under tensorflow/core/grappler/optimizers/arithmetic_optimizer_test.cc that asserts that the optimization is skipped for ArgMax(Relu(x)) and that the correct eager-mode index output is preserved when multiple inputs fall into the flat region:

```cpp
  TEST_F(ArithmeticOptimizerTest, OptimizeArgMaxOrArgMinOfMonotonicElementWiseDoNotChangeNonStrict) {
    tensorflow::Scope s = tensorflow::Scope::NewRootScope();
    
    // Inputs have multiple values in the flat region of Relu (x <= 0).
    // Relu(x) = [0.0f, 0.0f, 0.0f]
    // ArgMax(Relu(x)) evaluates to index 0 (first index of the maximum, 0.0f).
    // If rewritten to ArgMax(x), it would return index 1 (value -1.0f).
    const auto x = ops::Const(s.WithOpName("x"), {-2.0f, -1.0f, -3.0f}, {1, 3});
    Output relu = ops::Relu(s.WithOpName("relu"), x);
    Output arg_max = ops::ArgMax(s.WithOpName("arg_max"), relu, 1);
    Output final_out = ops::Identity(s.WithOpName("final_out"), arg_max);
  
    GrapplerItem item;
    item.fetch = {"final_out"};
    TF_CHECK_OK(s.ToGraphDef(&item.graph));
    const auto tensors_expected = EvaluateNodes(item.graph, item.fetch);
    ASSERT_EQ(tensors_expected.size(), 1);
  
    GraphDef output;
    ArithmeticOptimizer optimizer;
    EnableOnlyOptimizeMaxOrMinOfMonotonic(&optimizer);
    OptimizeAndPrune(&optimizer, &item, &output);
    const auto tensors = EvaluateNodes(output, item.fetch);
    ASSERT_EQ(tensors.size(), 1);
  
    // Assert that output indices match (correctness check)
    test::ExpectTensorEqual<int64_t>(tensors[0], tensors_expected[0]);
  
    // Expect that the Relu node remains in the graph (i.e. not optimized away)
    EXPECT_EQ(output.node_size(), item.graph.node_size());
    bool found_arg_max = false;
    for (const auto& node : output.node()) {
      if (node.name() == "arg_max") {
        EXPECT_EQ(node.input(0), "relu");
        found_arg_max = true;
      }
    }
    EXPECT_TRUE(found_arg_max);
  }
```

Once updated, we will import this into Google's internal repository, run our test suites, and merge it!
